data_structures/src/chain.rs

```diff
@@ -1939,7 +1939,98 @@ pub enum DataRequestStage {
     TALLY,
 }
 
-pub type UnspentOutputsPool = HashMap<OutputPointer, ValueTransferOutput>;
+/// Unspent Outputs Pool
+#[derive(Clone, Debug, Default, Eq, PartialEq, Serialize, Deserialize)]
+pub struct UnspentOutputsPool {
+    /// Map of output pointer to value transfer output
+    map: HashMap<OutputPointer, ValueTransferOutput>,
+    /// Map of transaction hash to a tuple of:
+    /// * The epoch of the block that included the transaction
+    /// * A reference count, used to keep this map clear after removing transactions
+    transaction_epoch: HashMap<Hash, (Epoch, u32)>,
+}
+
+impl UnspentOutputsPool {
+    pub fn get<Q: ?Sized>(&self, k: &Q) -> Option<&ValueTransferOutput>
+    where
+        OutputPointer: std::borrow::Borrow<Q>,
+        Q: std::hash::Hash + Eq,
+    {
+        self.map.get(k)
+    }
+
+    pub fn contains_key<Q: ?Sized>(&self, k: &Q) -> bool
+    where
+        OutputPointer: std::borrow::Borrow<Q>,
+        Q: std::hash::Hash + Eq,
+    {
+        self.map.contains_key(k)
+    }
+
+    pub fn insert(
+        &mut self,
+        k: OutputPointer,
+        v: ValueTransferOutput,
+        block_epoch: Epoch,
+    ) -> Option<ValueTransferOutput> {
+        let transaction_id = k.transaction_id;
+        let old_vto = self.map.insert(k, v);
+        if old_vto.is_none() {
+            let (transaction_epoch, refcount) = self
+                .transaction_epoch
+                .entry(transaction_id)
+                .or_insert((block_epoch, 0));
+            *refcount += 1;
+            // A transaction can only live inside one block, so the epoch must be equal
+            assert_eq!(block_epoch, *transaction_epoch);
+        } else {
+            // Tried to insert an existing transaction again
+            // TODO: This shouldn't really happen, panic?
+        }
+
+        old_vto
+    }
+
+    pub fn remove(&mut self, k: &OutputPointer) -> Option<ValueTransferOutput> {
+        let vto = self.map.remove(k);
+
+        if vto.is_some() {
+            // Decrease refcount
+            let (_epoch, refcount) = self.transaction_epoch.get_mut(&k.transaction_id).unwrap();
+            *refcount -= 1;
+            if *refcount == 0 {
+                self.transaction_epoch.remove(&k.transaction_id);
+            }
+        }
+
+        vto
+    }
+
+    pub fn drain<'a>(
+        &'a mut self,
+    ) -> impl Iterator<Item = (OutputPointer, ValueTransferOutput)> + 'a {
+        self.map.drain()
+    }
+
+    pub fn iter<'a>(
+        &'a self,
+    ) -> impl Iterator<Item = (&'a OutputPointer, &'a ValueTransferOutput)> + 'a {
+        self.map.iter()
+    }
+
+    /// Returns the epoch of the block that included the transaction referenced
+    /// by this OutputPointer. The difference between that epoch and the
+    /// current epoch is the "coin age".
```

Comment:
That description doesn't exactly match the specification:
> Entries of the UTXO set need to be tagged with a reference to the block in which they were created so as to track how "old" are they in terms of **how many blocks have been consolidated into the block chain since their creation**.

We are in time to discuss the merits of one option vs. the other.

Reply:
Any @witnet/devs want to chime in on this?

Reply:
I implemented both options so we don't lose much time after deciding the correct apporach.

The main arguments for each side are:

* Block number is consistent with the implementation of `extra_commit_rounds` and `extra_reveal_rounds`
* Epoch number can be used to calculate exactly when the UTXO will be available

Reply:
Just for the record, before resolving the conversation — the final implementation **DOES match** the specification as per https://github.com/witnet/WIPs/pull/18.